TESTS/mbed_hal/lp_us_tickers/ticker_api_tests.h

```diff
@@ -0,0 +1,129 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** \addtogroup hal_ticker_tests */
+/** @{*/
+
+#ifndef TICKER_API_TESTS_H
+#define TICKER_API_TESTS_H
+
+#include "device.h"
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Test that ticker_init can be called multiple times and
+ * ticker_init resets the internal count and disables the ticker interrupt.
```

Comment:
Hi
I don't understand this part...
Why the internal counter has to be reset ?

Reply:
I guess that after initialization we want to count from 0, but we should ask requirement author - @c1728p9 ?

Reply:
The intent is to fully reset the counter and having it start counting from 0. This is to prevent unexpected behavior if code outside your application initialized the timer. For example, if a buggy bootloader enables the timer and leaves it on when starting your application, your application needs to function the same regardless of what the bootloader did to the timer.

Reply:
I don't think I will reset the RTC time during LPT init...

Reply:
I agree with Jerome, I don't see the point here. On STM32 targets, we're using the same HW instance for implementing STM32 SDK internal tick and MBED us tickers. So I'd say that we don't want to reset the HW timer ... 
Also one naive question: when several tickers instances are created, wouldn't it break the running instance when creating a new one and resetting counter ?

Reply:
>Also one naive question: when several tickers instances are created, wouldn't it break the running instance when creating a new one and resetting counter ?

I assume that `xx_ticker_init()` function is for HAL usage only and should be called only once. In current version HAL ticker layer is implemented to call ticker init target specific driver only once:
https://github.com/ARMmbed/mbed-os/blob/67dfdde5628d355965e50cdc3e5c61e500cb5bf5/hal/mbed_ticker_api.c#L30-L36

>I don't think I will reset the RTC time during LPT init...

I understand that low power ticker is based on RTC. Init will be called only once probably during system startup. But if LPT is based on RTC, then what will happen with ticker schedule if user decide to change RTC time using `void set_time(time_t t)` function?

Reply:
ticker init - if RTC is involved, that one wont be reset if already running (this is already handled in the HAL implementations).

only one consumer of HAL ticker there is, init happens only once. Why we can't reset the counter for a high freq ticker during the first init, even if the peripheral is shared between drivers (in the STM32 SDK, this would be known to a user, that HAL takes ownership and does this reset once, and would reset the timer) ?

Reply:
The test description said:
"Test that ticker_init can be called multiple times and ticker_init resets the internal count and disables the ticker interrupt"
You're stating now that the init will be called only once but you want to test that it can be called multiple times, I'm not sure to understand ? 
About resetting the counter : how do you want to test that counter is 0 and why ? My possible fear is that the STM32 SDK has started the HW timer already during early init stage, when mbed ticker init is reached, the counter will not be 0 and there is no point in resetting to 0.  Also I don't understand the point of havint it to 0, how would that improve the system overall ?

Reply:
From my point of view if we assume that ticker_init will be called only once, then we can modify requirements and maybe add additional requirement: `Return without any action if ticker interface has been already initialised`. But this needs to be done by the requirements authors (@0xc0170, @c1728p9). 

If the decision is to change the requirement. I would like to do it in this PR and update test case at once.

Reply:
The point of these requirements is also to define some testable behaviours.
Is it really required to define and test an `init` function ?
It could be static and called internally "on demand".

`init` could even be removed from `ticker_interface_t` if there was not a requirement on disabling the ticker interrupt.

https://github.com/ARMmbed/mbed-os/blob/67dfdde5628d355965e50cdc3e5c61e500cb5bf5/hal/us_ticker_api.h#L64

Reply:
>  the requirement to reset the count can be relaxed

Does this need another PR ?

Reply:
Yes, this will be done in the next step. Please create an issue.